lib/cache/inventory_cache.go

```diff
@@ -0,0 +1,655 @@
+	ic.processEvents(ctx, watcher)
+
+	return nil
+}
+
+// waitForPrimaryCacheInit waits for the primary cache to be initialized.
+func (ic *InventoryCache) waitForPrimaryCacheInit(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return trace.Wrap(ctx.Err())
+	case <-ic.cfg.PrimaryCache.FirstTimeInit():
+		return nil
+	}
+}
+
+// setupWatcher sets up a watcher for instance and bot_instance events.
+func (ic *InventoryCache) setupWatcher(ctx context.Context) (types.Watcher, error) {
+	watcher, err := ic.cfg.Events.NewWatcher(ctx, types.Watch{
+		Name: "inventory_cache",
+		Kinds: []types.WatchKind{
+			{Kind: types.KindInstance},
+			{Kind: types.KindBotInstance},
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return watcher, nil
+}
+
+// waitForWatcherInit waits for the watcher to finish initializing.
+func (ic *InventoryCache) waitForWatcherInit(ctx context.Context, watcher types.Watcher) error {
+	select {
+	case <-ctx.Done():
+		// Context was cancelled
+		return trace.Wrap(ctx.Err())
+
+	case event := <-watcher.Events():
+		if event.Type != types.OpInit {
+			return trace.BadParameter("expected OpInit event, got %v", event.Type)
+		}
+		return nil
+	}
+}
+
+// populateCache reads teleport and bot instances and populates the cache with rate limiting.
+func (ic *InventoryCache) populateCache(ctx context.Context, readsPerSecond int) error {
+	limiter := rate.NewLimiter(rate.Limit(readsPerSecond), readsPerSecond)
+
+	if err := ic.populateInstances(ctx, limiter); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := ic.populateBotInstances(ctx, limiter); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// populateInstances reads teleport instances from the inventory service with rate limiting.
+func (ic *InventoryCache) populateInstances(ctx context.Context, limiter *rate.Limiter) error {
+	instanceStream := ic.cfg.Inventory.GetInstances(ctx, types.InstanceFilter{})
+
+	for instanceStream.Next() {
+		if err := limiter.Wait(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+
+		instance := instanceStream.Item()
+
+		instanceV1, ok := instance.(*types.InstanceV1)
+		if !ok {
+			ic.cfg.Logger.WarnContext(ctx, "Instance is not InstanceV1", "instance", instance.GetName())
+			continue
+		}
+
+		// Add it to the cache
+		ui := &inventoryInstance{instance: instanceV1}
+		if err := ic.store.put(ui); err != nil {
+			ic.cfg.Logger.WarnContext(ctx, "Failed to add instance to cache", "instance", instanceV1.GetName(), "error", err)
+			continue
+		}
+	}
+
+	return trace.Wrap(instanceStream.Done())
+}
+
+// populateBotInstances reads bot instances from the bot instance service with rate limiting.
+func (ic *InventoryCache) populateBotInstances(ctx context.Context, limiter *rate.Limiter) error {
+	var pageToken string
+
+	for {
+		if err := limiter.Wait(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+
+		botInstances, nextToken, err := ic.cfg.BotInstanceCache.ListBotInstances(
```

Comment:
If the watcher is from the backend we can't fetch the data from the cache: morally, the OpInit from the backend might refer to a "time" after the current "time" of the cache, so we might miss items that are not yet in the cache but that were already written in the backend so they won't appear in the event stream anymore (there's no single event order so it's a little more complicated than that, but for each individual backend item the reasoning holds).